### ISSUE TITLE
SOLVED: Firmware upgrade not working for dnt IPdio mini

### ISSUE THREAD

**Issue by rehand** [jisotalo/reciva-radio-patching#2]
Hello,
I tried the tutorial on my dnt IPdio mini but it seems that it doesn't upgrade the firmware.

When I invoke "Upgrade Firmware" it says "Upgrading Firmware" and then it reboots.

The Reciva Patcher says:
`http: got request.
http: failed to handle request
dns: query for www.reciva.com
dns: redirected to local http server.
dns: query for www.reciva.com
dns: redirected to local http server.
http: got request.
server: patch info requested.
dns: query for www.reciva.com
dns: redirected to local http server.
dns: query for www.reciva.com
dns: redirected to local http server.
dns: query for www.reciva.com
dns: redirected to local http server.
http: got request.
server: patch info requested.
dns: query for http:`

Do you have any idea how to get it on my radio working?

**Comment by jisotalo** [comment]
Have you tried multiple times? I think I had the same problem too. Just try again few times and it should start the update, hopefully

**Comment by rehand** [comment]
It seems that I've to press the "Update" button after pressing start. Now it updated and I've got the webserver. Then I tried to upgrade to sharpfin 2.0 as described but after updating I didn't get webserver and also no telnet access.

Now I'm trying to updating firmware again, hopefully it works again.

**Comment by jisotalo** [comment]
Good to hear! Hopefully it works. Never know what differences the radios have.

If it's not working you could try the old sharpfin version too

**Comment by rehand** [comment]
After patching the firmware again I don't get a webserver anymore...

Do you have any idea - may resetting the radio via menu change anything?

**Comment by jisotalo** [comment]
You can reset the radio and try again. Or just try firmware update again. Also taking the power cable out and back in might help.

**Comment by jisotalo** [comment]
Did you get it to work @rehand?

**Comment by rehand** [comment]
No sadly I didn't get it working until now. I already tried resetting the radio, flashing sharpfin-ep2.install directly as well as using the "Sharpfin-uninstall_0.1.patch" (from a different website).

Flashing always works but I don't get a connection via HTTP or telnet to the radio.

**Comment by crymsn** [comment]
Hello,
my trials fail at step 8. when the server demands 'Please check settings'. What did i miss? Server runs as administrator. I switched off my firewall. As far as i understand it, it has nothing to do with my iPdio mini, but with my laptop that is supposed to serve as, well, the server. Right? 
There is a file 'server.der.crt' in the same folder. Right-click context menu  shows the option 'install certificate'. Do i have to install it manually beforehand? Maybe, a silly question, but i'm not that familiar with hacking old internet radios. And maybe the problem has another cause.
Kind regards and thanks in advance.

**Comment by matpanel** [comment]
maybe you didn't enter the dns server address. It says 8.8.8.8 but it is empty in reality. 
You need to write it yourself, as it says in step 2...
Note: The textbox is empty at start (it only has a placeholder 8.8.8.8)

**Comment by matpanel** [comment]
> It seems that I've to press the "Update" button after pressing start. Now it updated and I've got the webserver. 

You should also press the update button after pressing start, as @rehand mentioned. 
Also, I have been told by peterkeet from the [iradioforum](https://iradioforum.net/forum/index.php) that the process could really be done in one step. As the sharpfin-ep2.install already contains the patch, one could select the patch url radio button (instead of the patch file) and paste the sharpfin-ep2.install url there. It will also not complain this way, if the link is https. [Link to the post](https://iradioforum.net/forum/index.php?topic=2969.msg25982#msg25982).

**Comment by crymsn** [comment]
ooh, ya, it was the 8.8.8.8 that was missing. 
Then, like rehand, i managed to install sharpfin to the radio, step 17, but the upgrade to vers. 2.0. failed. No more connection via web browser. Several attempts with factory resets and installing different versions of sharpfin including vers 2.0 directly via server.exe failed, more precisely: server-log and radio display always indicate it's installing, but no connection to the radio afterwards is possible, neither through web browser nor via telnet. And, the radio continues to look for the receiva gateway.
It was worth giving it a try, but it looks like my IPdio mini will experience some e-waste recycling in the near future.
Thanx again for your support.

**Comment by jisotalo** [comment]
Did you try with the older version? Perhaps it's enough for you. Not sure if you can edit presets from it?

Even with all patches, the radio will still complain about gateway. You have to press cancel/back button to access presets.

**Comment by matpanel** [comment]
It is strange... because you say that is seems to install successfully...
I would also make sure: 
- the clock is set up to the right time and date
- dhcp is set to auto

maybe do another reboot of the device? 
and I would double check the IP address (it could have changed)

**Comment by matpanel** [comment]
But it could also be the case that it just doesn't work on that particular device/firmware combination. 
Here is a [list of devices and compatible firmware versions](https://www.sharpfin.org/index.php?title=Radios).

**Comment by crymsn** [comment]
yes, dhcp ist set to auto, time and date are correct, IP-address checked and it is still the same. Firmware is version v257-a-865-a-162-c-002, which unfortunately is not in the list. And yes, i tried the elder version as well.

**Comment by rehand** [comment]
Why exactly is the clock relevant? I guess only for https, but also telnet connection to port 443 is refused, so "nobody" is listening on it. Or is there a different reason?

**Comment by jisotalo** [comment]
@crymsn: I'm not following.

You said you installed sharpfin to the radio. That means that after installing the version 1, you managed to connect it by browser before starting v2 update? Am I right? 

If so, could you stop there and not update v2. Instead, try the older version.

@rehand I have no idea if the clock is needed, haven't touched it myself. The telnet is only at port 23.

**Comment by rehand** [comment]
Yes I know that the telnet server itself listens only to port 23 (which also doesn't work), but if a webserver would be running a telnet connection to port 80 or 443 should be possible (instead of being refused).

I now have the same issue as @crymsn, after installing version 1 the connection via browser was possible, but after upgrading to v2 not anymore. After installing v1 again the browser connection still doesn't work - so it seems to be broken, no matter which version I'm flashing.

**Comment by jisotalo** [comment]
Damn. What happens if you do a factory reset from the radio? And after that, install the v1 again?

The connection isn't working without and with DHCP? So not even with static IP?

**Comment by rehand** [comment]
I also tried doing a factory reset but it also didn't work. And I tried with static IP and also with DHCP, with and without reboot.

It seems that something is broken due to the upgrade to v2.

I already tried the sharpfin uninstall script but it didn't help.

**Comment by matpanel** [comment]
> 
> 
> Why exactly is the clock relevant? I guess only for https, but also telnet connection to port 443 is refused, so "nobody" is listening on it. Or is there a different reason?

I just wanted to exclude all possible reasons... Time differences between machines can lead to connectivity issues with file shares and network protocols in general.

**Comment by crymsn** [comment]
@jisotalo: that's exactly what i did, before every new attempt i did a factory reset of the IPdio mini. And to be on the very safe side, i did it step by step one more time again right now, installed the sharpfin-base_0.3.patch, that once, and only once, allowed me to contact the radio through my browser. But, nope, it refuses and so does telnet. Well, ping works fine, the radio is there somewhere in my network and shows the exact time and date. DHCP or static IP do not make a difference.

**Comment by jisotalo** [comment]
@rehand @crymsn I'm sad to hear that. Hopefully we will find a way to fix that.

I'll added a note to the guide to stay at older version with IPdio mini radios.

**Comment by Cathesso** [comment]
I ran into the same problem as I've seen this github-project with the warning a bit too late. :'D
The radio itself still works with my saved favorites, just telnet and the webserver don't work anymore.

**Comment by asmnoak** [comment]
Hi there! It went well on my Tangent Quattro. I've restored my favorite presets. Thanks!

**Comment by hpn0t0ad** [comment]
Updating to `sharpfin-ep2.install` on my IPdio mini was a bit tricky since there was all but a single 1k block left on the rootfs after installing `sharpfin-base_0.3`. Unfortunately, ipkg needs some space for its database at /usr/lib/ipkg/. So before installing the ep2 I deleted some unnecessary (for me) locales from /usr/share/locale/ to free a few hundred kB on the rootfs. After that, the update went smoothly.

**Comment by rehand** [comment]
Does somebody know how to build a `.install` or `.patch` file which removes some files, e.g. all sharpfin files and maybe also some unused locales? This could help to get the IPdio mini working again

**Comment by jisotalo** [comment]
@rehand: You can open the `.install` file for example with 7zip, as it's an archive. Then you could modify the `install-me` file there to delete files.

However, I haven't touched those files myself and I don't have any spare unit to test/train with.

Something like
```
#!/bin/sh

(echo 1 > /dev/misc/S3C2410\ watchdog) >> /mnt/debug/patch.log 2>&1
./lcdprint "Patching" "Starting" >> /mnt/debug/patch.log 2>&1
sleep 1

mount / -orw,remount

#add file removal code here

mount / -oro,remount

echo 1 > /dev/misc/S3C2410\ watchdog
sync
sleep 2


./lcdprint "Patching Done" >> /mnt/debug/patch.log 2>&1
./lcdprint "Rebooting" >> /mnt/debug/patch.log 2>&1
reboot
```

Perhaps @hpn0t0ad has some idea?

**Comment by rehand** [comment]
That's promising, thanks for the hint. I'll try to create a patch file which also deletes some locales resp. a file to get the radio working again with the old version.

Could take a while as I have not much spare time a the moment but I'll post it here if I have success.

**Comment by Trashwarez** [comment]
Same issues here on my dnt IPdio mini. Since my tries to upgrade to v2 2 years ago all ports are closed.

**Comment by AndrewKosen** [comment]
Thank you so much for the detailed instructions on how to get the radio working again! I have a Tangent radio I did everything, everything works! But I couldn't finish the quick wifi wake up and preset expansion functions. I did everything correctly, made the appropriate changes to the config file, double-checked that the changes were saved, but the radio did not react in any way to the added lines.

**Comment by rehand** [comment]
Please create a separate issue as it seems to be a other problem

**Comment by ymienert** [comment]
@jisotalo at first: many thank for your great stuff and support - 4 reciva devices will have a long live now!

I can confirm the lines by hpn0t0ad - if I delete some files before I start the Update to release 2 is all well running. I use now this lines below to insert some removes at the "install-me" by the base installation file (sharpfin-base_0.3.patch - Starting at line 11):

mount / -orw,remount
#add file removal code here - let only 3 - 4 - 5 locales remain
#rm -rf /usr/share/locale/en_GB
#rm -rf /usr/share/locale/en_GB.revo
rm -rf /usr/share/locale/en_US
rm -rf /usr/share/locale/fr_FR
rm -rf /usr/share/locale/es_ES
rm -rf /usr/share/locale/ca_ES
#rm -rf /usr/share/locale/de_DE
#rm -rf /usr/share/locale/de_DE.OLY
rm -rf /usr/share/locale/da_DK
rm -rf /usr/share/locale/it_IT
rm -rf /usr/share/locale/nl_NL
rm -rf /usr/share/locale/pt_PT
rm -rf /usr/share/locale/fi_FI
rm -rf /usr/share/locale/sv_SE
rm -rf /usr/share/locale/no_NO
rm -rf /usr/share/locale/zh_CN
rm -rf /usr/share/locale/zh_TW
rm -rf /usr/share/locale/ja_JP

Please modify the remarks for your needed languages before.
If this lines included i get also already "bricked" IPdio's back on the ports and have enough space for the release 2. 

Also the big brother by the "IPdio mini" called "IPdio" (without mini) have exact the same issues with the low space on the drive. @jisotalo: may you could extend your fine installation guide with this fact?

**Comment by jisotalo** [comment]
That sounds awesome @ymienert! Very nice work.

Could you share the patch file, so others could try it? If it works for others too, I will surely update the guide. It would be easier for people to try it if it was a ready-to-use file, and then it should be 1:1.

**Comment by Trashwarez** [comment]
I can confirm that the approach of @ymienert works. My "bricked" dnt IPdio returned to life and reopened the ports. Attached please find the modified patch file. But be aware: **The only language settings left are english and german there**.
[sharpfin-base_0.3-MOD BY TRASHWARE.zip](https://github.com/jisotalo/reciva-radio-patching/files/7384680/sharpfin-base_0.3-MOD.BY.TRASHWARE.zip)

**Comment by jisotalo** [comment]
That's great news!

So the process is to install your patch (which seems to be 0.3 just with those deletion commands) and then continue normally and install Sharpfin 2?

Check it out @rehand & @crymsn!

**Comment by ymienert** [comment]
Yes, that's the way for me. (just 2 times tested on IPdio, 1 times on IPdio mini) If only 4 or 5 language settings are left on the drive is on the IPdio's and on the IPdio's mini enough space to install Sharpfin 2 like descripted at your guide.

**Comment by Trashwarez** [comment]
Yes, that's the way it works. There is even enough space left after the media server was installed.

![IPDio after patch](https://user-images.githubusercontent.com/68970598/138258153-1ff8cda0-4bd6-47e4-bacf-a1e9b4ae5e0c.png)

**Comment by Cathesso** [comment]
I patched my IPdio mini with the file @Trashwarez posted. That unbricked my radio. Then I installed Sharpfin 2.0 and added the streaming media server just as described and everything worked like a charm! :)

**Comment by crymsn** [comment]
Yeah! Many Thanx to @Trashwarez. Like @Cathesso i managed to patch, update to Sharpfin 2.0 and add the media server to my IPdio mini. Some stations i added to the presets work fine, others refuse to connect. Any idea why? The media server, i.e. 'Browse radio:', showed up only after i changed the language from Deutsch to English, and after changing back to Deutsch it's still there. 
It's good to have my IPdio mini back 🥇 👍 Thanx to all of you.

**Comment by jisotalo** [comment]
Thanks all for help! I updated the readme and added the IPdio patch file to repository.

@crymsn have you checked that those radio URLs work ok on your computer? And note that HTTPS urls won't work. You need to try if the radio supports HTTP. So like `https://domain.com/stream.mp3` -> `http://domain.com/stream.mp3`

I'll leave this issue open for now, so people will notice it :)

**Comment by rehand** [comment]
Thanks all, I used the script from @Trashwarez and it unbricked my radio and I was also able to update to 2.0 now :-)

Now I just need to find out how to browse my presents directly on the radio without the remote control.

**Comment by crymsn** [comment]
@jisotalo OK, that's it, they work when i simply delete the _s_ in http_s_ and make it http. 👍

**Comment by ymienert** [comment]
Due to the V2 issue already discussed I had a deeper look at the OS behind the IPdio mini. So i wrote a script to let the "reciva gateway not responding" message disappear and to get the station who is stored  at position one at the presets store, playing. This script is only tested on two IPdio mini yet. But I have no idea why it should not be working on another reciva radio device as well. Prerequisites are that a Sharpin V2 is installed and a reciva device including remote controll.(this script is calling the remote controll keys at the Sharpin web server) Disadvantages are that the alarm clock is running a bit unreliably (in this case you can only hear a beeping). Also that the backlight is flashing for some seconds a minute. @jisotalo: Would you like me to upload an ".install" with a modified "install-me"? Or is it wrong place with the wrong Issue?

**Comment by jisotalo** [comment]
Sounds great!  Perhaps you could create a new issue "Fix to gateway not responding" or similar. Feel free to add it as attachment there!

**Comment by ymienert** [comment]
Thank you, it's done.

**Comment by Stallone888** [comment]
I still have a problem with my IPdio Tune. I do all the steps, everything, double check, and still I have error:  Error Checking for Update Retry Later. I have tried everything, and nothing works. 
Does this patch work on Tune version of IPdio?
After I start Reciva pacher and restart the tuner I got some reports, but after that when I try Firmware, nothing, same error.

gui: trying to determine IP.
dns: server is running...
gui: servers started...
dns: query for ntp.reciva.com
dns: redirected to local http server.
dns: query for ntp.reciva.com
dns: redirected to local http server.
dns: query for ntp.reciva.com
dns: redirected to local http server.
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal1.7803986842.com
dns: redirected to local http server.
dns: query for portal1.7803986842.com
dns: redirected to local http server.
dns: query for portal2.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal2.7803986842.com
dns: query for portal3.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal3.7803986842.com
dns: query for portal4.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal4.7803986842.com
dns: query for portal5.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal5.7803986842.com
dns: query for portal6.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal6.7803986842.com
dns: query for portal7.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal7.7803986842.com
dns: query for portal8.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal8.7803986842.com
dns: query for portal9.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal9.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal10.7803986842.com
dns: query for portal10.7803986842.com
dns: query for portal11.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal11.7803986842.com
dns: query for portal12.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal12.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal13.7803986842.com
dns: query for p1.ext.h4.west.us.reciva.com
dns: redirected to local http server.
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal13.7803986842.com
dns: query for p1.ext.h4.west.us.reciva.com
dns: redirected to local http server.
dns: query for portal14.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for portal14.7803986842.com
dns: query for portal15.7803986842.com
dns: query for portal15.7803986842.com
dns: query for portal16.7803986842.com
dns: query for portal16.7803986842.com
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
gui: patch settings changed.
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for 5.1.168.192.in-addr.arpa
dns: query for www.reciva.com
dns: redirected to local http server.
dns: query for www.reciva.com
dns: redirected to local http server.


For whey I push Firmware update I got just this two lines:
dns: query for www.reciva.com
dns: redirected to local http server.

Am I doing things wrongly since IPdio is connected to swich on which are connected few devices.... I even tried disabling firewall but no luck

**Comment by mitch7676** [comment]
Thank you for working on this issue, it solved my problem with the dnt ipdio as well and saved it from the trash bin. Great spirit!

**Comment by Jerry2734** [comment]
Hi. It worked on my IPdio mini pro in general. Now I can edit the presets via web interface! But I get the "DNS Servers Invalid" messages on every turn on after patch (similar situation as before the patch). Same problem as described here for another hardware "https://github.com/jisotalo/reciva-radio-patching/issues/17". I tried to turn on the "keep wifi alive" option (wifi-standby-power-menu) via telnet, but no difference, i have to push the back button twice on every turn on of the radio. Do I have to live with this or is there any solution for this?

**Comment by VinayAnroedh** [comment]
I am sorry for responding to a post that was made a long time ago, but how can I edit the patch file for IPdio to leave English, German and Dutch? Both English and Dutch are important for me to remain on my iPdio Tune and when I try to open the patch file in notepad or notepad++ (Windows), it looks like a jumbled mess. Is there a different program I can use to edit it or is there some other way?

**Comment by rehand** [comment]
Hi,
you can open the patch file with 7-Zip and then edit the file install-me, there are the commands which delete the languages, so you have to remove "rm -rf /usr/share/locale/nl_NL" from there to keep dutch as well.

**Comment by VinayAnroedh** [comment]
Hi, I am sorry, I went ahead with the install after discussing it with my parents (who are going to be using the radio) and they preferred an english working radio over a dutch not working one. I loaded their presets in for them, so it will most likely be okay in English, however, if someone still has the Dutch files, I would still like to have them (together with the knowledge of how to add them back). I did try to open the file with WinRar, but a part of the file was still illegible, so I gave up.

**Comment by rehand** [comment]
It's already removed from my radio as well. Maybe somebody got 2.0 installed without having to remove unused language files and still has it or it can somehow be extracted from the original firmware?
